op-program/client/driver/driver.go

```diff
@@ -24,44 +24,98 @@ type Derivation interface {
 	Step(ctx context.Context) error
 }
 
-type EngineState interface {
+type Pipeline interface {
+	Step(ctx context.Context, pendingSafeHead eth.L2BlockRef) (outAttrib *derive.AttributesWithParent, outErr error)
+	ConfirmEngineReset()
+}
+
+type Engine interface {
 	SafeL2Head() eth.L2BlockRef
+	PendingSafeL2Head() eth.L2BlockRef
+	TryUpdateEngine(ctx context.Context) error
+	derive.ResetEngineControl
 }
 
 type L2Source interface {
 	derive.Engine
 	L2OutputRoot(uint64) (eth.Bytes32, error)
 }
 
-type NoopFinalizer struct{}
+type Deriver interface {
+	SafeL2Head() eth.L2BlockRef
+	SyncStep(ctx context.Context) error
+}
 
-func (n NoopFinalizer) OnDerivationL1End(ctx context.Context, derivedFrom eth.L1BlockRef) error {
-	return nil
+type MinimalSyncDeriver struct {
+	logger            log.Logger
+	pipeline          Pipeline
+	attributesHandler driver.AttributesHandler
+	l1Source          derive.L1Fetcher
+	l2Source          L2Source
+	engine            Engine
+	syncCfg           *sync.Config
+	initialResetDone  bool
+	cfg               *rollup.Config
 }
 
-func (n NoopFinalizer) PostProcessSafeL2(l2Safe eth.L2BlockRef, derivedFrom eth.L1BlockRef) {}
+func (d *MinimalSyncDeriver) SafeL2Head() eth.L2BlockRef {
+	return d.engine.SafeL2Head()
+}
+
+func (d *MinimalSyncDeriver) SyncStep(ctx context.Context) error {
+	if !d.initialResetDone {
+		if err := d.engine.TryUpdateEngine(ctx); !errors.Is(err, derive.ErrNoFCUNeeded) {
+			return err
+		}
+		if err := derive.ResetEngine(ctx, d.logger, d.cfg, d.engine, d.l1Source, d.l2Source, d.syncCfg, nil); err != nil {
+			return err
+		}
+		d.pipeline.ConfirmEngineReset()
+		d.initialResetDone = true
+	}
 
-func (n NoopFinalizer) Reset() {}
+	if err := d.engine.TryUpdateEngine(ctx); !errors.Is(err, derive.ErrNoFCUNeeded) {
+		return err
+	}
+	if err := d.attributesHandler.Proceed(ctx); err != io.EOF {
+		// EOF error means we can't process the next attributes. Then we should derive the next attributes.
+		return err
+	}
 
-var _ derive.FinalizerHooks = (*NoopFinalizer)(nil)
+	attrib, err := d.pipeline.Step(ctx, d.engine.PendingSafeL2Head())
+	if err != nil {
+		return err
+	}
+	d.attributesHandler.SetAttributes(attrib)
+	return nil
```

Comment:
To be equivalent with the previous `engine_queue` behaviour this should return `derive.NotEnoughData` which then gets op-program to not process additional blocks. It's actually reading one extra batch than it needs to (including in the old code) so I'd lean towards returning `nil` here and just checking if the safe head has advanced sufficiently after each step rather than requiring a `NotEnoughData` error to trigger it. Better to not be tightly coupled to particular semantics.